ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java

```diff
@@ -0,0 +1,619 @@
+    assertEquals(helixCluster.getPartitions().size(), (long) getGaugeValue("partitionReadWriteCount", Long.class));
+    assertEquals(0, (long) getGaugeValue("partitionSealedCount", Long.class));
+    assertEquals(helixCluster.getDiskCapacity(), (long) getGaugeValue("rawTotalCapacityBytes", Long.class));
+    assertFalse(getGaugeValue("isMajorityReplicasDownForAnyPartition", Boolean.class));
+    helixCluster.bringAllInstancesDown();
+    assertTrue(getGaugeValue("isMajorityReplicasDownForAnyPartition", Boolean.class));
+    if (useComposite) {
+      helixCluster.bringAllInstancesUp();
+      PartitionId partition = clusterManager.getWritablePartitionIds().get(0);
+      assertEquals(0, getCounterValue("getPartitionIdFromStreamMismatchCount"));
+
+      ReplicaId replicaId = partition.getReplicaIds().get(0);
+      assertEquals(0, getCounterValue("getReplicaIdsMismatchCount"));
+
+      // bring the replica down.
+      for (int i = 0; i < clusterMapConfig.clusterMapFixedTimeoutDiskErrorThreshold; i++) {
+        clusterManager.onReplicaEvent(replicaId, ReplicaEventType.Disk_Error);
+      }
+      clusterManager.getWritablePartitionIds();
+      assertEquals(0, getCounterValue("getPartitionIdFromStreamMismatchCount"));
+
+      InputStream partitionStream = new ByteBufferInputStream(ByteBuffer.wrap(partition.getBytes()));
+      clusterManager.getPartitionIdFromStream(partitionStream);
+      assertEquals(0, getCounterValue("getWritablePartitionIdsMismatchCount"));
+
+      clusterManager.hasDatacenter("invalid");
+      clusterManager.hasDatacenter(dcs[0]);
+      assertEquals(0, getCounterValue("hasDatacenterMismatchCount"));
+
+      DataNodeId dataNodeId = clusterManager.getDataNodeIds().get(0);
+      assertEquals(0, getCounterValue("getDataNodeIdsMismatchCount"));
+
+      clusterManager.getDataNodeId(dataNodeId.getHostname(), dataNodeId.getPort());
+      assertEquals(0, getCounterValue("getDataNodeIdMismatchCount"));
+    }
+  }
+
+  /**
+   * Test {@link AmbryDataNode}, {@link AmbryDisk}, {@link AmbryPartition} and {@link AmbryReplica}.
+   * @throws Exception
+   */
+  @Test
+  public void helixClusterManagerComponentsTest() throws Exception {
+    if (useComposite) {
+      return;
+    }
+    String hostName = "localhost";
+    int portNum = 2000;
+    long rackId = 1;
+    int sslPortNum = 3000;
+    Properties props = new Properties();
+    props.setProperty("clustermap.host.name", hostName);
+    props.setProperty("clustermap.cluster.name", "clusterName");
+    props.setProperty("clustermap.datacenter.name", "DC0");
+    props.setProperty("clustermap.ssl.enabled.datacenters", "DC1");
+    ClusterMapConfig clusterMapConfig1 = new ClusterMapConfig(new VerifiableProperties(props));
+    props.setProperty("clustermap.datacenter.name", "DC1");
+    ClusterMapConfig clusterMapConfig2 = new ClusterMapConfig(new VerifiableProperties(props));
+
+    // AmbryDataNode test
```

Comment:
The cluster components are better tested together - as they are interrelated - need to create `AmbryDataNode` for creating an `AmbryDisk`, for example. Having them all in a single method makes things easier.

Have moved this whole method to a test of its own, independent of Helix, as this is intended to be used with *any* dynamic cluster manager. Have also made `HelixClusterManagerCallback` and implementation of an interface called `ClusterManagerCallback`, which is what the dynamic components will use.